pkg/controllers/provisioning/v1alpha1/allocator/constraints.go

```diff
@@ -15,50 +15,129 @@ limitations under the License.
 package allocator
 
 import (
+	"context"
+	"crypto/md5"
+	"fmt"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
 	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/utils/scheduling"
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Constraints support a known set of labels
+const (
+	ArchitectureLabel       = "kubernetes.io/arch"
+	OperatingSystemLabelKey = "kubernetes.io/os"
+	ZoneLabel               = "topology.kubernetes.io/zone"
 )
 
-type Constraints struct{}
+type Constraints struct {
+	kubeClient client.Client
+}
 
-func (c *Constraints) Group(pods []*v1.Pod) []*cloudprovider.Constraints {
-	groups := []*cloudprovider.Constraints{}
+func (c *Constraints) Group(ctx context.Context, spec *v1alpha1.ProvisionerSpec, pods []*v1.Pod) ([]*cloudprovider.Constraints, error) {
+	// Groups uniqueness is tracked by md5(NodeConstraints)
+	groups := map[[md5.Size]byte]*cloudprovider.Constraints{}
```

Comment:
I wonder if it would be safer to use https://github.com/mitchellh/hashstructure instead of hand-rolling md5-based approach? I'm not aware of specific problem necessarily, just figuring there might be some weird edge case they handle?